src/enzyme_ad/jax/Dialect/Ops.cpp

```diff
@@ -123,6 +123,19 @@ Attribute JITCallOp::removeArgAttrsAttr() { return nullptr; }
 
 Attribute JITCallOp::removeResAttrsAttr() { return nullptr; }
 
+// Follows that of stablehlo.custom_call
+void JITCallOp::getEffects(
+    SmallVectorImpl<SideEffects::EffectInstance<MemoryEffects::Effect>>
+        &effects) {
+  auto hasSideEffect = getHasSideEffectAttr();
+  if (hasSideEffect && !hasSideEffect.getValue())
+    return;
+  effects.emplace_back(MemoryEffects::Allocate::get());
```

Comment:
we shouldn't do it like this (it otherwise says that the call always allocates/read/writes/reads). We probably want to recurse into the callee, if available and get the effects from within

Reply:
we may also want to add a custom folder as well

Reply:
I'm more worried about having a specification on the _caller_ that the callee has no side effects. I'd rather have the callee have an attribute that we can check here. We could then have an analysis that adds such attributes on all functions (reusing from Enzyme-MLIR work) in a topological way and avoid recursive/expensive jumping over the call graph.

Reply:
If has_side_effect is not specified, we can check the callee and get the memory effects from https://github.com/EnzymeAD/Enzyme-JAX/pull/876

Reply:
I think we should always check the callee. The dangerous situation I see that the caller says "no side effects, trust me bro" while being wrong. This may also happen on the function itself if the body changes without having the attributes updated, but at least is easier to debug (and we can have an expensive verifier).